File: testdata/fixtures/kotlin/safe/SqlParameterized.kt
package com.example.safe

import android.database.sqlite.SQLiteDatabase
import androidx.room.Dao
import androidx.room.Query

class SafeUserRepository(private val db: SQLiteDatabase) {

    // Safe: parameterized rawQuery with selection args
    fun findUser(name: String): User? {
        val cursor = db.rawQuery("SELECT * FROM users WHERE name = ?", arrayOf(name))
        return cursor.use { if (it.moveToFirst()) mapUser(it) else null }
    }

    // Safe: parameterized rawQuery with multiple args
    fun findUserByIdAndName(id: String, name: String): User? {
        val cursor = db.rawQuery(
            "SELECT * FROM users WHERE id = ? AND name = ?",
            arrayOf(id, name)
        )
        return cursor.use { if (it.moveToFirst()) mapUser(it) else null }
    }

    private fun mapUser(cursor: android.database.Cursor): User {
        return User(
            id = cursor.getInt(0),
            name = cursor.getString(1)
        )
    }
}

// Safe: Room DAO with parameterized queries
@Dao
interface UserDao {
    @Query("SELECT * FROM users WHERE name = :name")
    fun findByName(name: String): User

    @Query("SELECT * FROM users WHERE id = :id")
    fun findById(id: Int): User?
}

data class User(val id: Int, val name: String)

File: testdata/fixtures/lua/safe/sql_parameterized.lua
-- Safe Lua: Parameterized SQL queries

local _M = {}

-- SAFE: Using quote_sql_str for escaping
function _M.find_user(db, username)
    local escaped = ngx.quote_sql_str(username)
    local sql = "SELECT * FROM users WHERE username = " .. escaped
    return db:query(sql)
end

-- SAFE: Static SQL query
function _M.get_all_users(db)
    return db:query("SELECT id, name, email FROM users ORDER BY id")
end

-- SAFE: Using ndk.set_var for SQL escaping
function _M.search(db, term)
    local safe_term = ndk.set_var.set_quote_sql_str(term)
    local sql = "SELECT * FROM products WHERE name LIKE " .. safe_term
    return db:query(sql)
end

return _M
